addons/sale_timesheet/tests/test_sale_timesheet.py

```diff
@@ -524,6 +524,171 @@ def test_timesheet_invoice(self):
         self.assertTrue(so_line_deliver_task_project.invoice_lines)
         self.assertEqual(so_line_deliver_task_project.qty_invoiced, timesheet4.unit_amount)
 
+    def test_timesheet_invoice_after_refund(self):
+        """ Test creation of invoice for sale order with timesheeted line that were partially refunded
+
+            1) Create a sale order with a timesheeted service invoiced on delivery
+            2) Log a portion of the order
+            3) Invoice the order
+            4) Refund/Credit a portion of the order
+            5) Log more hours (don't finish order)
+            6) Invoice
+            7) Refund/Credit a portion of the order
+            8) Log the rest of the hours (finish the order)
+            9) Invoice
+        """
+        today = Date.context_today(self.env.user)
+        sale_order = self.env['sale.order'].create({
+            'partner_id': self.partner_a.id,
+            'partner_invoice_id': self.partner_a.id,
+            'partner_shipping_id': self.partner_a.id,
+            'pricelist_id': self.company_data['default_pricelist'].id,
+        })
+        so_line_deliver_global_project = self.env['sale.order.line'].create({
+            'name': self.product_delivery_timesheet2.name,
+            'product_id': self.product_delivery_timesheet2.id,
+            'product_uom_qty': 50,
+            'product_uom': self.product_delivery_timesheet2.uom_id.id,
+            'price_unit': self.product_delivery_timesheet2.list_price,
+            'order_id': sale_order.id,
+        })
+        so_line_deliver_task_project = self.env['sale.order.line'].create({
+            'name': self.product_delivery_timesheet3.name,
+            'product_id': self.product_delivery_timesheet3.id,
+            'product_uom_qty': 50,
+            'product_uom': self.product_delivery_timesheet3.uom_id.id,
+            'price_unit': self.product_delivery_timesheet3.list_price,
+            'order_id': sale_order.id,
+        })
```

Comment:
You can create in batch. 🙂 
```suggestion
        so_line_deliver_global_project, so_line_deliver_task_project = self.env['sale.order.line'].create([
            {
                'name': self.product_delivery_timesheet2.name,
                'product_id': self.product_delivery_timesheet2.id,
                'product_uom_qty': 50,
                'product_uom': self.product_delivery_timesheet2.uom_id.id,
                'price_unit': self.product_delivery_timesheet2.list_price,
                'order_id': sale_order.id,
            },
            {
                'name': self.product_delivery_timesheet3.name,
                'product_id': self.product_delivery_timesheet3.id,
                'product_uom_qty': 50,
                'product_uom': self.product_delivery_timesheet3.uom_id.id,
                'price_unit': self.product_delivery_timesheet3.list_price,
                'order_id': sale_order.id,
            },
        ])
```